### ISSUE TITLE
Title „Improper Data Filtering“ Misleading

### ISSUE THREAD

**Issue by thomaskonrad** [OWASP/API-Security#2]
I think the title „Improper Data Filtering“ is a bit misleading as it sounds a bit too much like input filtering.

How about „Excessive Object Property Exposure“?

By the way, I love the idea of this project. Great job!

**Comment by JonnySchnittger** [comment]
Perhaps "improper data exposure" or "inappropriate data leakage" might be more accurate?

**Comment by thomaskonrad** [comment]
I like „data“ as it is more generic than „object“. I‘d stick with „excessive“ though as „improper“ has a too wide meaning.

How about merging the two: „Excessive Data Exposure“?

**Comment by JonnySchnittger** [comment]
"Excessive Data Exposure" works for me

**Comment by 007divyachawla** [comment]
Could be Excessive Data Leakage or Unintended Data Leakage, because its not been done unintentionally.

**Comment by xycloops123** [comment]
Data leakage or Excessive data exposure sounds better.

On Thu, Jun 13, 2019 at 9:51 PM Divya <notifications@github.com> wrote:

> Could be Excessive Data Leakage or Unintended Data Leakage, because its
> not been done unintentionally.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_OWASP_API-2DSecurity_issues_2-3Femail-5Fsource-3Dnotifications-26email-5Ftoken-3DABYQHE6IJ3EQGVQ3MMRT6KDP2MPTNA5CNFSM4HWTA6XKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODXVWOBA-23issuecomment-2D501966596&d=DwMCaQ&c=slrrB7dE8n7gBJbeO0g-IQ&r=ak5EWD_xYqaVS7Q-mYJmjg&m=MjLZVyeR0KhoR6Lpn68UGZwC8qeKL1LjklQnCwMZmNs&s=Crd0Rai4H-R2kTAAR0WR9DqPaf45_KUsnVRzNvO5oqQ&e=>,
> or mute the thread
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_ABYQHE5BHDOAXOUPWGYOT5TP2MPTNANCNFSM4HWTA6XA&d=DwMCaQ&c=slrrB7dE8n7gBJbeO0g-IQ&r=ak5EWD_xYqaVS7Q-mYJmjg&m=MjLZVyeR0KhoR6Lpn68UGZwC8qeKL1LjklQnCwMZmNs&s=uJYHBir2h3a7cJYJlUkTwJjl7XuBNpBYFdo_gt7u4pU&e=>
> .
>
-- 
*Thanks And Regards*
*Gaurav Mukherjee*
*CS Graduate Student at NYU*
*(347) 658-3762*

**Comment by PauloASilva** [comment]
Since:

* "leak" is commonly associated with threat agents being able to subtract data from a third-party system by means of exploitation
* "Improper Data Filtering" refers to data/object properties unintentionally exposed by the development team

I'll vote for "**Excessive Data Exposure**".

**Comment by thomaskonrad** [comment]
Totally with you Paulo. „Excessive“ means „more than necessary“ which is exactly what we want to say.

**Comment by JonnySchnittger** [comment]
Looks like we have a winner then?

**Comment by ErezYalon** [comment]
I also vote for "Excessive Data Exposure".
Let's give it a few more days in the sun for community responses before making the change.

**Comment by jmanico** [comment]
What makes this different than an access control problem?

I think three top ten items that is essentially just an access control problem is... 

...excessive.

--
Jim Manico
@Manicode

**Comment by JonnySchnittger** [comment]
It's not just access control, if you can manipulate parameters to cause undesired behavior e.g. a database exception the application may then expose internal information such as table names, server address etc...

**Comment by thomaskonrad** [comment]
It‘s quite different, because the root cause is not a missing access control, but a missing mechanism to specify which properties of an Entity (or object) should be serialized and included in the response.

In a very broad sense yes, it is a form of access control, but it does not fit an insecure direct object reference or a missing function-level access control. It is whitelisting of Entity properties to be serialized.

**Comment by thomaskonrad** [comment]
@JonnySchnittger, not sure if we are talking about the same thing here. The vulnerability is, as far as I understand, not about provoking an exception that includes sensitive data.

It is about whether the API (1) includes sensitive fields in the response by design (such as the password of a user when getting user details via, e.g., `/api/users/{userId}`, or (2) lets the client specify which fields to return (e.g., via GraphQL) and allows adding sensitive fields which the client should never receive.

**Comment by JonnySchnittger** [comment]
My assumption is the error is returned via the API, and should have filtered out e.g. StackTrace or Debug information removed (or not enabled in the 1st place)

**Comment by PauloASilva** [comment]
> @JonnySchnittger, not sure if we are talking about the same thing here. The vulnerability is, as far as I understand, not about provoking an exception that includes sensitive data.
> It is about whether the API (1) includes sensitive fields in the response by design (such as the password of a user when getting user details via, e.g., `/api/users/{userId}`, or (2) lets the client specify which fields to return (e.g., via GraphQL) and allows adding sensitive fields which the client should never receive.

You're right.
"Sensitive" isn't/shouldn't be part of category's name because it is business specific.

> @thomaskonrad My assumption is the error is returned via the API, and should have filtered out e.g. StackTrace or Debug information removed (or not enabled in the 1st place)

StackTraces and debug information are covered in [A7:2019 Security Misconfiguration][1]

[1]: https://github.com/OWASP/API-Security/blob/develop/2019/en/src/0xa7-security-misconfiguration.md

**Comment by JonnySchnittger** [comment]
@PauloASilva I accept that A7 covers them, perhaps it was a bad example :)

**Comment by pleothaud** [comment]
I also vote for "Excessive Data Exposure". 

Regarding Jim Manico's statement about this item overlapping with other Access Control related items (A1 and A5), I've got the impression that the first type of "Improper Data Filtering" (Client Side Data Filtering) is definitey not an Access Control problem, but the second one (Filter Manipulation) is IMHO covered by A5 as it is a lack of Access Control on a function call. 

WDYT?

**Comment by jmanico** [comment]
Train your Developers on Secure Coding

View this email in your browser (https://mailchi.mp/875ff264e773/manicode-security-june-2019?e=3a214cbd0c)


** Manicode Security
------------------------------------------------------------


** Secure Coding Education
June 2019 Update
------------------------------------------------------------
Secure Coding Classes
Thank you for being a customer, partner or associate of Manicode Security.

Manicode Security is now offering over a dozen different application security classes delivered by four different instructors.
* Ron Perris is now teaching classes in React, Node.js, Express and JavaScript secure coding.
* Jimmy Mesta is teaching classes in Kubernetes, AWS, GPC, Cloud Security and DevOps security.
* Dr. Philppe De Ryck is teaching classes in Advanced Web Security.
* And Jim Manico is teaching classes in a wide variety of Web and API Security topics.
* Classes are also available in Mobile Secure Coding and Penetration Testing.

You can download our full catalog here https://manicode.us19.list-manage.com/track/click?u=3b283d304057d06aaba6dc8fb&id=f5e5c7f0ff&e=3a214cbd0c

Please drop us a line at info@manicode.com (mailto:info@manicode.com?subject=Info-June-2019) if you have any questions about our application security classes!
Manicode Security is Partnered with Bit Discovery

The one singular thing literally every information security professional agrees upon is that all organizations must have an inventory of all their internet-connected assets. This includes all websites, name servers, mail servers, IoT devices, blogs, and so on. Of course, an inventory is also required by most every information security standard. And the reason is simple, you cannot secure what you don’t know that you own. Period. Creating a complete asset inventory has proven exceptionally technically challenging, time-consuming, and labor intensive. This has to change. Bit Discovery’s mission, having invested the last several years creating a fantastic product, is making sure every organization has a complete asset inventory. Instantly created. Automatically updated.

Please visit https://manicode.us19.list-manage.com/track/click?u=3b283d304057d06aaba6dc8fb&id=a6b3d01345&e=3a214cbd0c for more information or email contact@bitdiscovery.com (mailto:contact@bitdiscovery.com)  to schedule a live demo to create an asset inventory for your organization and receive a free trial account.

============================================================

This email was sent to reply@reply.github.com (mailto:reply@reply.github.com)
why did I get this? (https://manicode.us19.list-manage.com/about?u=3b283d304057d06aaba6dc8fb&id=94b725b0f5&e=3a214cbd0c&c=b62483214f)     unsubscribe from this list (https://manicode.us19.list-manage.com/unsubscribe?u=3b283d304057d06aaba6dc8fb&id=94b725b0f5&e=3a214cbd0c&c=b62483214f)     update subscription preferences (https://manicode.us19.list-manage.com/profile?u=3b283d304057d06aaba6dc8fb&id=94b725b0f5&e=3a214cbd0c)
Manicode Security . PO Box 676 . Anahola, HI 96703-0676 . USA

Email Marketing Powered by Mailchimp
http://www.mailchimp.com/monkey-rewards/?utm_source=freemium_newsletter&utm_medium=email&utm_campaign=monkey_rewards&aid=3b283d304057d06aaba6dc8fb&afl=1

**Comment by inonshk** [comment]
> 
> What makes this different than an access control problem? I think three top ten items that is essentially just an access control problem is... ...excessive.
> […](#)
> -- Jim Manico @manicode
> On Jun 14, 2019, at 4:36 AM, Erez Yalon ***@***.***> wrote: I also vote for "Excessive Data Exposure". Let's give it a few more days in the sun for community responses before making the change. — You are receiving this because you are subscribed to this thread. Reply to this email directly, view it on GitHub, or mute the thread.

Thanks for your feedback.
You're right, we could call it an access control problem.
Anyhow, access control is more of an idea rather than a specific mechanism. 
By reviewing hundreds of reports and input from many pentesters and security engineers, we found that access control problems are extremely common in APIs.
If the broken access control problems were less severe / less prevalent, we could put them under one unified category and explain, with a few words, about each one of the occurrences in a generic way (similar to A7).
But - and it's a big but, access control problems are the most common vector to attack APIs. we do not want to underestimate them, because of the simple fact this is the main flow of how companies getting breached today.

After spending a lot of time thinking about that, we decided to create the categories, because of the following differentiations:

**A1 - Broken Object Level Access Control**
What does the mechanism do?
- Validates that one user has access only to objects he should have access to. Objects can be users, receipts, shopping carts, etc..
Where does the mechanism exist?
- The Object Level Access Control mechanism is usually implemented in the code level, in every controller that receives object ID/s from the clients (e.g.: user_id, receipt_id, etc..)
It should validate that the logged-in user has access to perform the requested action on the specific object that is represented by the ID that had been sent from the client.
There are various ways to implement it, but a common approach is basically to call the authorization mechanism after/before every access to the ORM that is based on an ID that had been sent from the client.
A good example to express the idea is the manual of cancancan (a great authorization gem for Ruby On Rails):
`def show
  @post = Post.find(params[:id])
  authorize! :read, @post
end`
https://github.com/CanCanCommunity/cancancan#31-authorizations

**A5 - Missing Function/Level Access Control**

What does the mechanism do?
- Validates that the logged-in user has access to perform a specific action in the app.
It is usually based on the user's group \ role (admin, supervisor, regular user, guest, etc).
Why is it different than A1?
- Very simple, because it does not depend on an object ID. Let's say that we have an administrative endpoint, "DELETE /api/users/717" - with missing function-level access control. This endpoint receives an ID of an object (user 717), but it's not the issue. The API should always block access to this endpoint by regular users (that are not admins), regardless of the IDs that might have been sent.
Where does the mechanism exist?
- Usually, the mechanism is more abstracted. It can be implemented in the code (parent controller), configuration or even a separated component like an API gateway.

**A3 - Excessive Data Exposure**
In this case, it's not about a lack of mechanism. It's about the fact that the developers wrote a code that exposes more information than required.
Two common reasons for that:
1. The developers use a generic "to_json" that was implemented in the model itself / provided by the ORM.
2. The developers use a function from the model that had been written for other purposes.

Hope things make more sense now.
